### ISSUE TITLE
Point force kernel integration 

### ISSUE THREAD

**Issue by mallickrishg** [brendanjmeade/bemcs#18]
Just leaving a note here for tasks at hand

1. MATLAB expressions for linearly varying force along a line segment need to be converted to Python. They are derived for [antiplane](https://github.com/mallickrishg/testing_greensfunctions/blob/main/antiplane/LinForceKernelFS.m) and [plane strain](https://github.com/mallickrishg/testing_greensfunctions/blob/main/planestrain/LinForceKernelFS.m) separately.
2. Abandoning analytical solution for point force integral over triangles. We will instead use a combination of `scipy.dblquad()` for evaluations inside the source domain (and perhaps near-field too?) and `quadpy` for far-field evaluations. The use of dblquad is shown [here](https://github.com/brendanjmeade/bemcs/blob/main/point_source/point_source_experiments.ipynb) while the use of quadpy is shown [here](https://github.com/brendanjmeade/bemcs/blob/main/point_source/kelvin_point_source.ipynb).

**Comment by brendanjmeade** [comment]
@mallickrishg I'm writing this for myself to get organized regarding integrating point sources.  Thoughts on the last three bullet points?

#### Triangle integration notes
- Near field:
	- slow and high accuracy with [`scipy.integrate.dblquad`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.dblquad.html)  
	- [notebook](https://github.com/brendanjmeade/bemcs/blob/main/point_source/point_source_experiments.ipynb)

- Far-field:
	- fast and good enough accuracy with [`quadpy`](https://github.com/sigma-py/quadpy)
	- [notebook](https://github.com/brendanjmeade/bemcs/blob/main/point_source/kelvin_point_source.ipynb)

- Steps
- [ ] Near and far-field examples into a single clean notebook.
- [ ] Single right triangle first so that not rotations, translations or splitting are required.
- [ ] Plot residual for near and far-field over a grid to get a sense of where fast integration looses accuracy.
  - [ ] How does this depend on dimensions?
  - [ ] How does this depend on `dblquad` parameters?
  - [ ] Can we define an effective length scale for the transition between solvers?

- [ ] Integration over an arbitrary triangle.  Note: `quadpy` does this.  `dblquad` requires an arbitrary triangle to be split into two right triangles, and then each of these must be translated and rotated to the origin with one leg of the triangle alighted with $y=0$.
  - [ ] Create a geometric splitting and translation function.
  - [ ] Set up translation and rotation of observation coordinates.

- [ ] Sum over multiple triangles
  - [ ] Coarse mesh of a box.
    - [] Evaluate at quadratic BEM nodes.
  - [ ] Course mesh of a triangle.
    - [ ] Evaluate at quadratic BEM nodes.
  - [ ] Course mesh of something geologic (e.g., crude subduction zone)
    - [ ] Evaluate at quadratic BEM nodes.

- When will we actually need accurate near-field integration?
  - For the BEM solve phase only for triangles that touch a BEM element?
  - Is there an evaluation (with a known BEM solution) to consider?

**Comment by mallickrishg** [comment]
> * Integration over an arbitrary triangle.  Note: `quadpy` does this.  `dblquad` requires an arbitrary triangle to be split into two right triangles, and then each of these must be translated and rotated to the origin with one leg of the triangle alighted with y=0.

`dblquad` does not require us to split into 2 right triangles, it can deal with an arbitrary triangle. However, 1 of its sides need to be aligned with the y-axis. The 3 points for the triangle are $(0,0),(x_2,y_2),(0,y_3)$. So the only task is to rotate a given triangle such that one of its sides is $(0,0) \to (0,y_3)$



> When will we actually need accurate near-field integration?
> 
> * For the BEM solve phase only for triangles that touch a BEM element?

We need the near-field integration always. For the case of elastic heterogeneity, each body force triangle source results in non-zero near-field strains (perhaps even at $r=0$), and the body force strength is basically $f = \frac{\epsilon}{\mu}\frac{\partial\mu}{\partial x}$. The nice thing about the stresses from triangle body force is that it is smooth everywhere - so even when the triangle shares a side with or touches a slip element, there shouldn't be any issue. 

> * Is there an evaluation (with a known BEM solution) to consider?

We could attempt to match the analytical solutions for a 2-layer medium from [Rybicki](https://pubs.geoscienceworld.org/ssa/bssa/article/61/1/79/116912/The-elastic-residual-field-of-a-very-long-strike). Of course, we would not match it exactly with the triangle body forces - but we can basically get very close depending on how narrow  the region of non-zero $\frac{1}{\mu}\frac{\partial\mu}{\partial x}$ we permit to approximate the step change in $\mu$.

**Comment by brendanjmeade** [comment]
> `dblquad` does not require us to split into 2 right triangles, it can deal with an arbitrary triangle. However, 1 of its sides need to be aligned with the y-axis.

Nice!

> We need the near-field integration always. For the case of elastic heterogeneity, each body force triangle source results in non-zero near-field strains (perhaps even at $r=0$), and the body force strength is basically $f = \frac{\epsilon}{\mu}\frac{\partial\mu}{\partial x}$. The nice thing about the stresses from triangle body force is that it is smooth everywhere - so even when the triangle shares a side with or touches a slip element, there shouldn't be any issue. 

This is a part I still need to learn.  For the triangle case do we need the strain rate at the triangle centers or boundaries?  I would have thought centers...which may or may not be far-field.

**Comment by mallickrishg** [comment]
> This is a part I still need to learn. For the triangle case do we need the strain rate at the triangle centers or boundaries? I would have thought centers...which may or may not be far-field.

In any case I think we only evaluate displacements and displacement gradients/strains resulting from a body force at the center of a triangle. The only case I can think of evaluating $u_i,\frac{\partial u_i}{\partial x_j}$ at the triangle boundary is at the midpoint of a domain line element that is coincident with the triangle boundary.

For the heterogeneous material case, we don't really need to worry about neighbors since every triangle can be deal with independently. However, I don't know the details of how we will set up the problem for viscoelasticity.

**Comment by brendanjmeade** [comment]
> In any case I think we only evaluate displacements and displacement gradients/strains resulting from a body force at the center of a triangle. The only case I can think of evaluating $u_i,\frac{\partial u_i}{\partial x_j}$ at the triangle boundary is at the midpoint of a domain line element that is coincident with the triangle boundary.

I largely agree.  One modification is that we'll have to do the evaluation at all three quadratic nodes, which is definitely a near-field calculation.  I'm not sure if the triangle centroids will be mid-field or far-field.  Either way, if we just need the values at centroids the cost of the near-field integral should be cheap enough.

> For the heterogeneous material case, we don't really need to worry about neighbors since every triangle can be deal with independently. However, I don't know the details of how we will set up the problem for viscoelasticity.

One step at a time!

**Comment by brendanjmeade** [comment]
A first look (for me) at residuals between integration methods.  An empirical cut-off for switching between fast and slow methods will definitely be possible!
<img width="462" alt="Screenshot 2024-05-11 at 12 35 29 PM" src="https://github.com/brendanjmeade/bemcs/assets/4225359/c9b4198b-5c0c-47b1-b6f0-9e9d7b525e03">

**Comment by brendanjmeade** [comment]
Showing 3 component stress residuals [commit](https://github.com/brendanjmeade/bemcs/commit/b9824ccfa7cd01bba40161b7e65f638200f2670c)

<img width="825" alt="Screenshot 2024-05-14 at 3 49 59 PM" src="https://github.com/brendanjmeade/bemcs/assets/4225359/6481b0a5-5502-4f24-b467-813511c5eca8">

**Comment by mallickrishg** [comment]
Nice! Looks like something as simple as 2 or 3 times the circumcircle radius reduces the residuals down to ~0. Perhaps we don't even need shapely?

**Comment by brendanjmeade** [comment]
Circles look cooler but shapely is fun!  I'll mess around.

**Comment by mallickrishg** [comment]
Visually it looks like there is no notable difference in the shape of the residual for any of the stress components. Is this also the case when you apply the other component of the body force? Presumably, this is the response to $f_x$?

**Comment by brendanjmeade** [comment]
Seems so.  The attached figure is a $f_y = 1$ test.  Still a bunch of little things to do so that wrappers for integration method are identical.  This is a really fun mini project!

<img width="818" alt="Screenshot 2024-05-14 at 5 58 45 PM" src="https://github.com/brendanjmeade/bemcs/assets/4225359/f3cdc1f7-1c34-4638-a8ed-9577e56914c4">

**Comment by brendanjmeade** [comment]
Using @mallickrishg 's idea to just use a circle around the triangle (rather than the more interesting, quirky, dependency requiring, and weird shapely buffering approach) it looks like if simply pick a radius of convergence equal three times length from the triangle centroid to it's farthest vertex then we are pretty much guaranteed machine precision.  I haven't done anything exhaustive in terms of geometry experiments, but rather just tried five or six different geometries.  Thoughts?

<img width="359" alt="Screenshot 2024-05-14 at 10 32 09 PM" src="https://github.com/brendanjmeade/bemcs/assets/4225359/4aeb60a0-f41b-4d4f-a321-fee120b243b0">

**Comment by mallickrishg** [comment]
Yes! This figure is exactly what I had in mind. Seems like 2*d is sufficiently far away for quadrature to work.

**Comment by brendanjmeade** [comment]
I'm going with $3 \times \max d_\mathrm{t}$ and this figure so that we are well past any bend.  This will make the cutoff completely defensible.  Hopefully this figure stands the test of time.

I'll start working on a notebook that implements a nice functionalized version that allows for simply passing the triangle geometry, forcing, material properties and observation points and then simple returns the stresses and displacements using the appropriate method based on the triangle size and observation coordinates.  I still have to get translations and rotations working correctly.  I've got a few bugs somewhere.
